### PR TITLE
UCP/RMA: Implement blocking RMA/AMO using non-blocking APIs

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7471/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7474/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@hoopoepg @brminich pls take a look

File: src/ucp/rma/amo_basic.c [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+    req->send.buffer                = buffer;
+}
+
+static UCS_F_ALWAYS_INLINE
+void ucp_amo_init_post(ucp_request_t *req, ucp_ep_h ep, uct_atomic_op_t op,
+                       size_t op_size, uint64_t remote_addr, ucp_rkey_h rkey,
+                       uint64_t value)
+{
+    ucp_amo_init_common(req, ep, op, remote_addr, rkey, value, op_size);
+    req->send.uct.func              = ucp_amo_basic_progress_post;
```
Comment (hoopoepg) [inline src/ucp/rma/amo_basic.c]: extra spaces?

File: src/ucp/rma/amo_basic.c [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+                   ucp_rkey_h rkey, uint64_t value)
+{
+    ucp_amo_init_common(req, ep, op, remote_addr, rkey, value, op_size);
+    req->send.state.uct_comp.count  = 1;
+    req->send.state.uct_comp.func   = ucp_amo_completed_single;
+    req->send.uct.func              = ucp_amo_basic_progress_fetch;
+    req->send.buffer                = buffer;
+}
+
+static UCS_F_ALWAYS_INLINE
```
Comment (brminich) [inline src/ucp/rma/amo_basic.c]: minor: better use common formatting (i. e. `void` on the same lane)

File: src/ucp/rma/amo_basic.c (hunk 1 above)
Comment (brminich) [inline src/ucp/rma/amo_basic.c]: maybe no need to keep `=` alignment with other funcs?

File: src/ucp/rma/amo_basic.c [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+
+    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
+
+    op     = ucp_uct_op_table[opcode];
+    status = UCP_RKEY_RESOLVE(rkey, ep, amo);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    req = ucp_request_get(ep->worker);
```
Comment (brminich) [inline src/ucp/rma/amo_basic.c]: why do we need a request if it is not exposed in API?

File: src/ucp/rma/rma_basic.c [hunk 4]
```diff
@@ -420,8 +301,9 @@ ucs_status_t ucp_put_nbi(ucp_ep_h ep, const void *buffer, size_t length,
     }
 
     rma_config = &ucp_ep_config(ep)->rma[rkey->cache.rma_lane];
-    status = ucp_rma_nonblocking(ep, buffer, length, remote_addr, rkey,
-                                 ucp_progress_put, rma_config->put_zcopy_thresh);
+    status = ucp_rma_nonblocking(ep, (void*)buffer, length, remote_addr, rkey,
```
Comment (hoopoepg) [inline src/ucp/rma/rma_basic.c]: unaligned '='

File: src/ucp/rma/amo_basic.c [hunk 5]
```diff
@@ -0,0 +1,351 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2015.  ALL RIGHTS RESERVED.
```
Comment (hoopoepg) [inline src/ucp/rma/amo_basic.c]: 2018

File: src/ucp/rma/amo_basic.c [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+{
+    return ucp_atomic_fetch_b(ep, UCP_ATOMIC_FETCH_OP_FADD, add, result,
+                              sizeof(add), remote_addr, rkey, "atomic_fadd32");
+}
+
+UCS_PROFILE_FUNC(ucs_status_t, ucp_atomic_fadd64, (ep, add, remote_addr, rkey, result),
+                 ucp_ep_h ep, uint64_t add, uint64_t remote_addr, ucp_rkey_h rkey,
+                 uint64_t *result)
+{
+    return ucp_rma_wait(ep->worker,
```
Comment (brminich) [inline src/ucp/rma/amo_basic.c]: why not `ucp_atomic_fetch_b`?

File: src/ucp/rma/amo_basic.c [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+                                            result, sizeof(add), remote_addr,
+                                            rkey, (void*)ucs_empty_function),
+                        "atomic_fadd64");
+}
+
+UCS_PROFILE_FUNC(ucs_status_t, ucp_atomic_swap32, (ep, swap, remote_addr, rkey, result),
+                 ucp_ep_h ep, uint32_t swap, uint64_t remote_addr, ucp_rkey_h rkey,
+                 uint32_t *result)
+{
+    return ucp_rma_wait(ep->worker,
```
Comment (brminich) [inline src/ucp/rma/amo_basic.c]: why not `ucp_atomic_fetch_b`?

File: src/ucp/rma/amo_basic.c [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+                                            result, sizeof(swap), remote_addr,
+                                            rkey, (void*)ucs_empty_function),
+                        "atomic_swap32");
+}
+
+UCS_PROFILE_FUNC(ucs_status_t, ucp_atomic_swap64, (ep, swap, remote_addr, rkey, result),
+                 ucp_ep_h ep, uint64_t swap, uint64_t remote_addr, ucp_rkey_h rkey,
+                 uint64_t *result)
+{
+    return ucp_rma_wait(ep->worker,
```
Comment (brminich) [inline src/ucp/rma/amo_basic.c]: why not `ucp_atomic_fetch_b`?

File: src/ucp/rma/amo_basic.c [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+                                            rkey, (void*)ucs_empty_function),
+                        "atomic_swap64");
+}
+
+UCS_PROFILE_FUNC(ucs_status_t, ucp_atomic_cswap32,
+                 (ep, compare, swap, remote_addr, rkey, result),
+                 ucp_ep_h ep, uint32_t compare, uint32_t swap,
+                 uint64_t remote_addr, ucp_rkey_h rkey, uint32_t *result)
+{
+    ucs_status_t status;
```
Comment (brminich) [inline src/ucp/rma/amo_basic.c]: minor: can use the same inlined code with the func below

File: src/ucp/rma/amo_basic.c (hunk 3 above)
Comment (yosefe) [reply src/ucp/rma/amo_basic.c]: we need something to progress (same as with ucp_put_nbi, for example)

**Pull request merged**